### DISCUSSION TITLE
How to submit a new nemo action?

### DISCUSSION THREAD

**Discussion by goebbe** [linuxmint/discussions#635] [Questions] [answered]
I'm currently looking for a guide on "howto submit a nemo action" for newbies.  :-)  

Situation: I created a new "action" on my home computer - and I would like to submit this action. 
While I submitted bug reports in the past, I have no experience / knowledge about the common workflow - and I didn't find a detailed (step-by-step) description. 

The following comment by @rcalixte is the only place where some steps are described (as far as I understand):
https://github.com/orgs/linuxmint/discussions/372#discussioncomment-8949861

Could somebody check if the following is correct: 
1. First, one has to **fork the whole repository** with all the nemo-spice-actions: https://github.com/linuxmint/cinnamon-spices-actions/tree/master
2. Create a new Branch for the new action - from my fork
3. Add all the new files to my fork/branch
4. Send a pull request from my fork to the original repository 

Is it possible to do all this on github - or do I have to install git and pull / push from my computer to my fork? 

Sorry for all my questions. I would be great if somebody would comment/ answer my questions.

**Comment by rcalixte** [comment, marked as answer]
> Is it possible to do all this on github - or do I have to install git and pull / push from my computer to my fork?

Your process is correct and is the typical path for contributing on GitHub in general. GitHub has thorough documention for [getting started](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/getting-started/about-collaborative-development-models) with that if you are unfamiliar.

Ultimately, you can structure/upload however you feel is comfortable but it is highly recommended to have `git` installed locally because there are helper scripts in the root of the repository that you can only run on your machine, not from the GitHub web interface. These scripts range from creating the translation template to local testing of your Action to validating that the structure of your Action is correct. The goal is to accelerate the process and reduce the friction, especially for first-time contributors. Take a look at some of the details in the [README](https://github.com/linuxmint/cinnamon-spices-actions?tab=readme-ov-file#file-structure) as well.

There is a process where you can do this without `git` installed but it would be more cumbersome because you would need to download the repository copy manually, make your changes locally, and then upload them manually to the proper GitHub location. It is doable but keep in mind it is a path where you may not find much assistance due to the non-standardized approach.

**Reply by goebbe** [reply]
@rcalixte Thank you very much, your answer is very helpful! 
- Of course it is much easier to run the **helper scripts** (that are part of the repository), if I have my own, local copy of the repository (i.e. on my computer) - this makes a lot of sense.
- I still don't quite understand why forking and branching is necessary. Why not just clone / change / pull request - not sure if this makes any sense.   Instead, one does:  fork / branch/ clone / change / push / pullrequest     Is this just for better/ cleaner organisation? Or this about ownership/ rights? 
- Guess I should do my reading now, thanks for the pointers!

**Reply by rcalixte** [reply]
> * I still don't quite understand why forking and branching is necessary.

This is because you don't have write access to the project repository. The branch creation requires write permissions that you won't have for a repository unless you own it or are granted those permissions. Forking it grants you ownership permission of a clone and allows you to send those changes back to the source if you want. You can also clean up your fork once you are done making your changes and re-clone in the future or keep it around for more immediate changes.
